### PR TITLE
feat(avoidance): check if the avoidance path is in drivable area

### PR DESCRIPTION
**Pull request by satoshi-ota**
### <samp>🤖 Generated by Copilot at 3d133af</samp>

This pull request refactors and improves the avoidance module of the behavior path planner. It removes unused or redundant code, adds new variables and parameters to store and control the drivable area and the lane departure, and enhances the generation and validation of shift lines for obstacle avoidance. These changes aim to increase the safety and efficiency of the avoidance behavior.

This PR related to https://github.com/autowarefoundation/autoware_launch/pull/584.

---

```c++
  // check if the vehicle is in road.
  {
    const auto minimum_distance = 0.5 * planner_data_->parameters.vehicle_width +
                                  parameters_->hard_road_shoulder_margin -
                                  parameters_->max_departure_from_lane;

    const size_t start_idx = shift_lines.front().start_idx;
    const size_t end_idx = shift_lines.back().end_idx;

    for (size_t i = start_idx; i <= end_idx; ++i) {
      const auto p = getPoint(shifter_for_validate.getReferencePath().points.at(i));
      lanelet::BasicPoint2d basic_point{p.x, p.y};

      const auto shift_length = proposed_shift_path.shift_length.at(i);
      const auto bound = shift_length > 0.0 ? avoid_data_.left_bound : avoid_data_.right_bound;
      const auto THRESHOLD = minimum_distance + std::abs(shift_length);

      if (boost::geometry::distance(basic_point, lanelet::utils::to2D(bound)) < THRESHOLD) {
        RCLCPP_DEBUG_THROTTLE(getLogger(), *clock_, 1000, "road shoulder distance is not enough.");
        return false;
      }
    }
  }
```

## Tests performed

- [x] [PASS TIER IV INTENAL SCENARIOS](https://evaluation.tier4.jp/evaluation/reports/7a8e4a82-93eb-5548-8e5c-6dce992b4dd8?project_id=prd_jt)

## Effects on system behavior

Nothing.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 1]
```diff
@@ -2350,6 +2376,30 @@ bool AvoidanceModule::isValidShiftLine(
     }
   }
 
+  // check if the vehicle is in road.
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2379]:
```suggestion
  // check if the vehicle is in road. (yaw angle is not considered)
```

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 2]
```diff
@@ -2350,6 +2376,30 @@ bool AvoidanceModule::isValidShiftLine(
...
+    for (size_t i = start_idx; i <= end_idx; ++i) {
+      const auto p = getPoint(shifter_for_validate.getReferencePath().points.at(i));
+      lanelet::BasicPoint2d basic_point{p.x, p.y};
+
+      const auto shift_length = proposed_shift_path.shift_length.at(i);
+      const auto bound = shift_length > 0.0 ? avoid_data_.left_bound : avoid_data_.right_bound;
+      const auto THRESHOLD = minimum_distance + std::abs(shift_length);
+
+      if (boost::geometry::distance(basic_point, lanelet::utils::to2D(bound)) < THRESHOLD) {
+        RCLCPP_DEBUG_THROTTLE(getLogger(), *clock_, 1000, "road shoulder distance is not enough.");
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2397]: It seems this debug info is outputted even if road shoulder distance is big enough?
In my understanding, this condition is met when road shoulder distance is not enough or `max_departure_from_lane` is big?

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM!

File: planning/behavior_path_planner/include/behavior_path_planner/utils/avoidance/avoidance_module_data.hpp [hunk 3]
```diff
@@ -235,6 +235,9 @@ struct AvoidanceParameters
   // Even if the obstacle is very large, it will not avoid more than this length for left direction
   double max_left_shift_length{0.0};
 
+  // Validate vehicle departure from driving lane.
+  double max_departure_from_lane{0.0};
```
[the remark this answers is not in the document]
Comment (satoshi-ota) [reply planning/behavior_path_planner/include/behavior_path_planner/utils/avoidance/avoidance_module_data.hpp]: Thanks for your idea. Fixed.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 2 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp]: update comment!
```
following avoidance path that is generated from latest new shift line may cause deviation from drivable area.
```

**Pull request merged**